### PR TITLE
Daemons

### PR DESCRIPTION
**Pull request by andorp**
* Creates `Daemon` module
 * Add simple message passing based logging to Action
 * Use timeout daemon in persistent implementation

### CODE REVIEW

**Comment by fisx** [comment]
What do you mean by "timeout deamon for persistent implementation"?

File: src/Action/Implementation.hs [hunk 1]
```diff
@@ -57,7 +57,7 @@ instance HasSendMail ActionExcept ActionEnv Action where
     sendMailToAddress addr msg = MkAction $ sendMailToAddressIO addr msg
 
 instance ActionLog Action where
-    logEvent = actionIO . print
+    logEvent msg = actionIO =<< views envLogger ($ msg)
```
Comment (fisx) [inline src/Action/Implementation.hs]: heh.  very cool.

File: src/Daemon.hs [hunk 2]
```diff
@@ -0,0 +1,87 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+{-# OPTIONS_GHC -Werror -Wall #-}
+
+module Daemon
```
Comment (fisx) [inline src/Daemon.hs]: call this `LogDeamon`?

Comment (fisx) [inline src/Daemon.hs]: (or `MsgDeamon`, for consistency?)

Comment (fisx) [inline src/Daemon.hs]: ah.  there are more daemons. never mind.

**Comment by fisx** [comment]
@Mikolaj was planning to work on phase timeouts in #402.  Please make sure that you two don't get in each other's way.  (@andorp, Perhaps you can do only the logging for now, as planned?)

File: src/Persistent/Implementation/AcidState.hs [hunk 3]
```diff
@@ -50,21 +51,18 @@ mkRunPersistOnDisk cfg =
...
-            threadDelay delay_us
+        let checkpoint = do
+                logger cfg "[create acid-state checkpoint, archive]"
+                createCheckpoint st
+                createArchive st
+        let logException (SomeException e) = do
+                logger cfg ("error creating checkpoint or archiving changelog: " <> show e)
 
+        let deamon = timeoutDaemon (logger cfg) "checkpoint" delay_us checkpoint logException
+        tid <- deamon ^. timeoutDaemonStart
```
Comment (fisx) [inline src/Persistent/Implementation/AcidState.hs]: ok, this answers my earlier question.  thanks!  (-:

Comment (fisx) [inline src/Persistent/Implementation/AcidState.hs]: You're not making use of the `Daemon` class here, though.  The reason for that may be that this class is not particularly useful.  I might have missed something (I have before), but you should either come up with a reason for it (at least a potential, future one) or remove it.

Comment (andorp) [inline src/Persistent/Implementation/AcidState.hs]: It is used, I replaced `timeoutDaemonStart` etc with `start`.

Comment (fisx) [inline src/Persistent/Implementation/AcidState.hs]: > It is used, I replaced `timeoutDaemonStart` etc with `start`.

ok, good.  (i think i was reading a partially outdated version of the code.)

File: src/Daemon.hs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    return $ MessageDaemon loop sendMsg
+
+timeoutDaemon
+    :: SystemLogger
+    -> String
+    -> Int
+    -> (IO ())
+    -> (SomeException -> IO ()) -> Daemon ()
+timeoutDaemon logger name delay_us computation handleException = TimeoutDeamon $ do
```
Comment (fisx) [inline src/Daemon.hs]: It would be nice to make `delay_us` a `(delay_sec :: Double)`.  Feel free to disregard.

Comment (fisx) [inline src/Daemon.hs]: in fact, what about this:

```haskell
data Timespan =
    TimespanUs Int
    TimespanMs Int
    TimespanSec Int
    TimespanHourse Int
    TimespanDays Int
```

(with the correct syntax)

Comment (fisx) [inline src/Daemon.hs]: obviously i meant `hoarse`  (-:

**Comment by Mikolaj** [comment]
@andorp, @fisx: I think the actual daemons and the particulars of freezing and thawing are quite orthogonal, so I hope we can do that in parallel; I will just use the daemons to chase some more data structures that there are currently.

**Comment by andorp** [comment]
@Mikolaj That's great.

**Comment by fisx** [comment]
Doesn't look like it from the number of commits I made, but I like it!  Good to merge if you and travis agree.

Thanks!

**Pull request merged**